dispatcher/coaps.go

```diff
@@ -0,0 +1,128 @@
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+package dispatcher
+
+import (
+	. "github.com/openthread/ot-ns/types"
+	"github.com/simonlingoogle/go-simplelogger"
+)
+
+type CoapType int
+
+const (
+	CoapTypeConfirmable    CoapType = 0
+	CoapTypeNonConfirmable CoapType = 1
+	CoapTypeAcknowledgment CoapType = 2
+	CoapTypeReset          CoapType = 3
+)
+
+type CoapCode int
+
+type CoapMessageRecvInfo struct {
+	Timestamp uint64 `yaml:"time"`
+	DstNode   NodeId `yaml:"dst"`
+	SrcAddr   string `yaml:"src_addr"`
+	SrcPort   int    `yaml:"src_port"`
+}
+
+type CoapMessage struct {
+	Timestamp uint64                `yaml:"time"`
+	SrcNode   NodeId                `yaml:"src"`
+	ID        int                   `yaml:"id"`
+	Type      CoapType              `yaml:"type"`
+	Code      CoapCode              `yaml:"code"`
+	URI       string                `yaml:"uri,omitempty"`
+	DstAddr   string                `yaml:"dst_addr,flow"`
+	DstPort   int                   `yaml:"dst_port,flow"`
+	Error     string                `yaml:"error,omitempty"`
+	Receivers []CoapMessageRecvInfo `yaml:"receivers,flow"`
+}
+
+type coapsHandler struct {
+	messages []*CoapMessage
+}
+
+func (coaps *coapsHandler) OnSend(curTime uint64, nodeId NodeId, messageId int, coapType CoapType, coapCode CoapCode, uri string, peerAddr string, peerPort int) {
+	coaps.messages = append(coaps.messages, &CoapMessage{
+		Timestamp: curTime,
+		SrcNode:   nodeId,
+		ID:        messageId,
+		Type:      coapType,
+		Code:      coapCode,
+		URI:       uri,
+		DstAddr:   peerAddr,
+		DstPort:   peerPort,
+	})
+}
+
+func (coaps *coapsHandler) OnRecv(curTime uint64, nodeId NodeId, messageId int, coapType CoapType, coapCode CoapCode, uri string, peerAddr string, peerPort int) {
+	msg := coaps.findMessage(messageId, coapType, coapCode, uri)
+	if msg == nil {
+		simplelogger.Warnf("COAP message %d,%d,%d,%s not sent but received by Node %d", messageId, coapType, coapCode, uri, nodeId)
+		return
+	}
+
+	msg.Receivers = append(msg.Receivers, CoapMessageRecvInfo{
+		Timestamp: curTime,
+		DstNode:   nodeId,
+		SrcAddr:   peerAddr,
+		SrcPort:   peerPort,
+	})
+}
+
+func (coaps *coapsHandler) OnSendError(nodeId NodeId, messageId int, coapType CoapType, coapCode CoapCode, uri string, peerAddr string, peerPort int, error string) {
+	msg := coaps.findMessage(messageId, coapType, coapCode, uri)
+	if msg == nil {
+		simplelogger.Warnf("COAP message %d,%d,%d,%s not sent but received by Node %d", messageId, coapType, coapCode, uri, nodeId)
+		return
+	}
+
+	msg.Error = error
+}
+
+func (coaps *coapsHandler) findMessage(id int, coapType CoapType, code CoapCode, uri string) *CoapMessage {
+	for i := len(coaps.messages) - 1; i >= 0; i-- {
```

Comment:
Why search the message in reverse?

Reply:
For better performance. If a coap message is just received, it's very likely the message should be just sent, thus at the bottom of the message list.